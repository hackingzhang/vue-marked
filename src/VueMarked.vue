<script lang="ts">
import "./styles/github-markdown.css";

export default {
  name: "VueMarked",
};
</script>
<script lang="ts" setup>
import DOMPurify from "dompurify";
import { Marked, type MarkedExtension } from "marked";
import { computed, onUnmounted, ref, watch } from "vue";

import type { ColorScheme, PrefersColorScheme } from "./types";

type Props = {
  /** color schema: light, dark or auto, default to auto */
  colorScheme?: ColorScheme;
  /** Markdown content */
  markdown: string;
  /** If sanitize the output HTML to prevent XSS attach, default to true */
  sanitize?: boolean;
  /** marked option, @see https://marked.js.org/using_advanced#options */
  options?: MarkedExtension;
  /** marked extentions, @see https://marked.js.org/using_advanced#extensions */
  extentions?: Array<MarkedExtension>;
};
const props = withDefaults(defineProps<Props>(), {
  colorScheme: "auto",
  sanitize: true,
  options: () => ({}),
  extentions: () => [],
});

const marked = new Marked(props.options, ...props.extentions);

const html = ref("");
watch(
  () => props.markdown,
  async () => {
    const rawHTML = await marked.parse(props.markdown);
    if (props.sanitize) {
      html.value = DOMPurify.sanitize(rawHTML);
    } else {
      html.value = rawHTML;
    }
  },
  { immediate: true }
);

const perfersColorScheme = ref<PrefersColorScheme>("light");
const colorScheme = computed(() => {
  if (props.colorScheme === "auto") {
    return perfersColorScheme.value;
  } else {
    return props.colorScheme;
  }
});
const darkModeListener = (ev: MediaQueryListEvent) => {
  console.log(ev);
  perfersColorScheme.value = ev.matches ? "dark" : "light";
};
const darkModeQuery = window.matchMedia("(prefers-color-scheme: dark)");
perfersColorScheme.value = darkModeQuery.matches ? "dark" : "light";
darkModeQuery.addEventListener("change", darkModeListener);
onUnmounted(() => {
  darkModeQuery.removeEventListener("change", darkModeListener);
});

defineExpose({
  marked: marked,
});
</script>

<template>
  <div v-html="html" :data-theme="colorScheme" class="v-marked"></div>
</template>

<style>
.v-marked {
  /* prevent katex styles from disrupting the layout */
  position: relative;
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
}
</style>
