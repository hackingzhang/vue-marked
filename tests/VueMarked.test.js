import { mount } from "@vue/test-utils";
import markedKatex from "marked-katex-extension";
import { describe, expect, test } from "vitest";

import VueMarked from "../src/VueMarked.vue";
import mdContent from "../playground/default-markdown.ts";

/**
 * Wait in milliseconds
 * @param {number} milliseconds
 * @returns {Promise}
 */
function sleep(milliseconds) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, milliseconds);
  });
}

describe("VueMarked.vue", () => {
  test("Should render empty content", () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown: "",
      },
    });

    expect(wrapper.text().length).toBe(0);
  });

  test("Should render markdown content", async () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown: mdContent,
      },
    });

    await sleep(100);
    expect(wrapper.html()).matchSnapshot();
  });

  test("Should sanitize output HTML by default", async () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown:
          "<p>paragraph will exist</p>\n<script>console.log('script tag will be sanitized')</script>",
      },
    });

    await sleep(100);
    const resultWrapper = wrapper.find("script");
    expect(resultWrapper.exists()).toBe(false);
  });

  test("Should NOT sanitize output HTML if specified", async () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown:
          "<p>paragraph will exist</p>\n<script>console.log('script tag will be sanitized')</script>",
        sanitize: false,
      },
    });

    await sleep(100);
    const resultWrapper = wrapper.get("script");
    expect(resultWrapper.exists()).toBe(true);
  });

  test("options prop should work", async () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown: "line1\nline2",
        options: {
          breaks: true,
        },
      },
    });

    await sleep(0);
    const resultWrapper = wrapper.get("br");
    expect(resultWrapper.exists()).toBe(true);
  });

  test("extentions prop should work", async () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown: `Inline: $E=mc^2$

Block:
$$
c = \\pm\\sqrt{a^2 + b^2}
$$`,
        extentions: [
          markedKatex({
            throwOnError: false,
          }),
        ],
      },
    });

    await sleep(0);
    expect(wrapper.html()).matchSnapshot();
  });

  test("colorScheme prop should work", () => {
    const wrapper = mount(VueMarked, {
      props: {
        markdown: "",
        colorScheme: "dark",
      },
    });

    expect(wrapper.attributes()["data-theme"]).toBe("dark");
  });
});
