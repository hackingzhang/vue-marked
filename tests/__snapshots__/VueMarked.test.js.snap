// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VueMarked.vue > Should render markdown content 1`] = `
"<div data-theme="light" class="v-marked">
  <h1>h1 Heading</h1>
  <h2>h2 Heading</h2>
  <h3>h3 Heading</h3>
  <h4>h4 Heading</h4>
  <h5>h5 Heading</h5>
  <h6>h6 Heading</h6>
  <h2>Horizontal Rules</h2>
  <hr>
  <hr>
  <hr>
  <h2>Emphasis</h2>
  <p><strong>This is bold text</strong></p>
  <p><strong>This is bold text</strong></p>
  <p><em>This is italic text</em></p>
  <p><em>This is italic text</em></p>
  <p><del>Strikethrough</del></p>
  <h2>Blockquotes</h2>
  <blockquote>
    <p>Blockquotes can also be nested...</p>
    <blockquote>
      <p>...by using additional greater-than signs right next to each other...</p>
      <blockquote>
        <p>...or with spaces between arrows.</p>
      </blockquote>
    </blockquote>
  </blockquote>
  <h2>Lists</h2>
  <p>Unordered</p>
  <ul>
    <li>Create a list by starting a line with <code>+</code>, <code>-</code>, or <code>*</code></li>
    <li>Sub-lists are made by indenting 2 spaces:<ul>
        <li>Marker character change forces new list start:<ul>
            <li>Ac tristique libero volutpat at</li>
          </ul>
          <ul>
            <li>Facilisis in pretium nisl aliquet</li>
          </ul>
          <ul>
            <li>Nulla volutpat aliquam velit</li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Very easy!</li>
  </ul>
  <p>Ordered</p>
  <ol>
    <li>
      <p>Lorem ipsum dolor sit amet</p>
    </li>
    <li>
      <p>Consectetur adipiscing elit</p>
    </li>
    <li>
      <p>Integer molestie lorem at massa</p>
    </li>
    <li>
      <p>You can use sequential numbers...</p>
    </li>
    <li>
      <p>...or keep all the numbers as <code>1.</code></p>
    </li>
  </ol>
  <p>Start numbering with offset:</p>
  <ol start="57">
    <li>foo</li>
    <li>bar</li>
  </ol>
  <h2>Code</h2>
  <p>Inline <code>code</code></p>
  <p>Indented code</p>
  <pre><code>// Some comments
line 1 of code
line 2 of code
line 3 of code
</code></pre>
  <p>Block code "fences"</p>
  <pre><code>Sample text here...
</code></pre>
  <p>Syntax highlighting using <a href="https://github.com/markedjs/marked-highlight">highlight extention</a></p>
  <pre><code class="language-javascript">var foo = function (bar) {
  return bar++;
};

console.log(foo(5));
</code></pre>
  <h2>Tables</h2>
  <table>
    <thead>
      <tr>
        <th>Option</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>data</td>
        <td>path to data files to supply the data that will be passed into templates.</td>
      </tr>
      <tr>
        <td>engine</td>
        <td>engine to be used for processing templates. Handlebars is the default.</td>
      </tr>
      <tr>
        <td>ext</td>
        <td>extension to be used for dest files.</td>
      </tr>
    </tbody>
  </table>
  <p>Right aligned columns</p>
  <table>
    <thead>
      <tr>
        <th align="right">Option</th>
        <th align="right">Description</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td align="right">data</td>
        <td align="right">path to data files to supply the data that will be passed into templates.</td>
      </tr>
      <tr>
        <td align="right">engine</td>
        <td align="right">engine to be used for processing templates. Handlebars is the default.</td>
      </tr>
      <tr>
        <td align="right">ext</td>
        <td align="right">extension to be used for dest files.</td>
      </tr>
    </tbody>
  </table>
  <h2>Links</h2>
  <p><a href="http://dev.nodeca.com">link text</a></p>
  <p><a href="http://nodeca.github.io/pica/demo/" title="title text!">link with title</a></p>
  <p>Autoconverted link <a href="https://github.com/nodeca/pica">https://github.com/nodeca/pica</a> (enable linkify to see)</p>
  <h2>Images</h2>
  <p><img src="https://octodex.github.com/images/minion.png" alt="Minion">
    <img src="https://octodex.github.com/images/stormtroopocat.jpg" alt="Stormtroopocat" title="The Stormtroopocat">
  </p>
  <p>Like links, Images also have a footnote style syntax</p>
  <p><img src="https://octodex.github.com/images/dojocat.jpg" alt="Alt text" title="The Dojocat"></p>
  <p>With a reference later in the document defining the URL location:</p>
  <h2>Task List</h2>
  <ul>
    <li><input checked="" disabled="" type="checkbox"> Support extentions</li>
    <li><input checked="" disabled="" type="checkbox"> Support options</li>
    <li><input checked="" disabled="" type="checkbox"> Enable light/dark theme</li>
    <li><input disabled="" type="checkbox"> Add examples</li>
  </ul>
  <h2>HTML</h2>
  <p style="color: red;">This is a <span style="color: white; background-color: green;">&lt;p&gt;</span> element in HTML</p>

  <h2>Katex</h2>
  <p>Implement with <a href="https://github.com/UziTech/marked-katex-extension">marked-katex-extension</a></p>
  <p>This is inline katex: $c = \\pm\\sqrt{a^2 + b^2}$</p>
  <p>This is block level katex:</p>
  <p>$$
    c = \\pm\\sqrt{a^2 + b^2}
    $$</p>
</div>"
`;

exports[`VueMarked.vue > extentions prop should work 1`] = `
"<div data-theme="light" class="v-marked">
  <p>Inline: <span class="katex"><span class="katex-mathml"></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.05764em;">E</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.8141em;"></span><span class="mord mathnormal">m</span><span class="mord"><span class="mord mathnormal">c</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></p>
  <p>Block:
    $$
    c = \\pm\\sqrt{a^2 + b^2}
    $$</p>
</div>"
`;
